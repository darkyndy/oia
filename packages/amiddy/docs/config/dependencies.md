# deps
Array that has one or more objects where each object represents a dependency.
Every dependency should have patters that will resolve.


## dependency.ip

- Data type: String
- Required: Yes if [dependency.name](#dependencyname) does not have value
- Default value: N/A
- Example value: `127.1.2.3`
- Details:
    - IP for the dependency.


## dependency.name

- Data type: String
- Required: Yes if [dependency.ip](#dependencyip) does not have value
- Default value: N/A
- Example value:
    - `domain.com`
    - `sub.domain.com`
- Details:
    - Host name for the dependency.
    - If [dependency.ip](#dependencyip) is set this value is ignored


## dependency.https

- Data type: Boolean
- Required: No
- Default value: `false`
- Example value:
    - `true`
    - `false`
- Details:
    - Specify if the dependency uses secure protocol.


## dependency.port

- Data type: Number
- Required: No
- Default value:
    - `80` if [dependency.https](#dependencyhttps) has falsy value
    - `443` if [dependency.https](#dependencyhttps) has `true` as value
- Example value:
    - `3000`
    - `8080`
- Details:
    - Specify port for the dependency.


### dependency.patterns

- Data type: Array<String>
- Required: Yes
- Default value: N/A
- Example value:
    - ```json
      [
        "/api/**"
      ]
      ```
    - ```json
      [
        "/test/**",
        "/api/*",
        "/logout"
      ]
      ```
- Details:
    - Specify patterns that if they match will use proxy the request to this dependency.
    - For the pattern testing [micromatch.isMatch](https://www.npmjs.com/package/micromatch#ismatch) is used having as options `{contains: true}`


### dependency.mocks

- Data type: Array<Object>
- Required: No
- Default value: N/A
- Example value:
    - ```json
      [
        {
          "patterns": ["/info**"],
          "response": {"backup":  "if we fail to get fixture content"},
          "status": 206,
          "fixture": "__local__/info.txt"
        }
      ]
      ```
    - ```json
      [
        {
          "methods": ["GET", "POST"],
          "patterns": ["/exact-path", "/user/**"],
          "response": {"notes": "for specified patterns at GET or POST respond with this and status code 200"}
        }
      ]
      ```
- Details:
    - Specify mocks for this dependency.


#### dependency.mocks.patterns

- Data type: Array<String>
- Required: Yes
- Default value: N/A
- Example value:
    - ```text/plain
      "patterns": ["/exact-path", "/user/**"],
      ```
    - ```text/plain
      "patterns": ["/info**"],
      ```
- Details:
    - Specify the pattern for request url to mock.
    - For the pattern testing [micromatch.isMatch](https://www.npmjs.com/package/micromatch#ismatch) is used having as options `{contains: true}`


#### dependency.mocks.methods

- Data type: Array<String>
- Required: No
- Default value: N/A
- Example value:
    - ```text/plain
      "methods": ["GET", "POST"],
      ```
    - ```text/plain
      "methods": ["PATCH"],
      ```
- Details:
    - Specify what method or methods should be mocked for the patterns.
    - Accept all methods if it is not specified.


#### dependency.mocks.status

- Data type: Any
- Required: No
- Default value: 200
- Example value:
    - ```text/plain
      "status": 200,
      ```
    - ```text/plain
      "status": 404,
      ```
- Details:
    - Specify mocked response status code.
    - Returns status code 200 when is not specified.


#### dependency.mocks.response

- Data type: Any
- Required: No
- Default value: '' (empty string)
- Example value:
    - ```text/plain
      "response": {"prop": "json response"},
      ```
    - ```text/plain
      "response": "text response",
      ```
    - ```text/plain
      "response": "{prop: 'json response as string'}",
      ```
- Details:
    - Specify mocked response body.


#### dependency.mocks.fixture

- Data type: String
- Required: No
- Default value: N/A
- Example value:
    - ```text/plain
      "fixture": "relative/path/to/file",
      ```
- Details:
    - Specify the relative path to the file that contains the response body.
    - If file content can be read it overwrites the value from `response`.


#### dependency.mocks.headers

- Data type: Object
- Required: No
- Default value: N/A
- Example value:
    - ```text/plain
      "headers": {
        "X-header": "this header will appear on the response",
        "X-header-mock": "also this header will appear on the response",
      },
      ```
- Details:
    - Headers that should be added to the response.
    - If the same header also exist on `proxy.response.headers` the value from here will be used.


