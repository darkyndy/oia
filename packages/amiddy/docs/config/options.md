
# options
Amiddy options.


## options.recorder
Options for recorder.


### options.recorder.enabled
- Data type: Boolean
- Required: No
- Default value: false
- Details:
    - Specify if you want to record every response from dependencies


### options.recorder.fileNamePattern
- Data type: String
- Required: No
- Default value: '{METHOD}-{PATH}.{EXT}'
- Details:
    - File name based on tokens where to save the response.
    - Available tokens are: `METHOD`, `PATH`, `EXT` and `STATUS`
    - To determine extension [npm mime-types](https://www.npmjs.com/package/mime-types) is used


### options.recorder.ignorePatterns
- Data type: Array<String>
- Required: No
- Default value: []
- Example value:
    - ```json
      [
        "/api/**"
      ]
      ```
    - ```json
      [
        "/test/**",
        "/api/*",
        "/logout"
      ]
      ```
- Details:
    - Specify patterns that if they match will not save the response.
    - For the pattern testing [micromatch.isMatch](https://www.npmjs.com/package/micromatch#ismatch) is used having as options `{contains: true}`


### options.recorder.path
- Data type: String
- Required: No
- Default value: '__amiddy__/records'
- Details:
    - Folder where to save responses.
