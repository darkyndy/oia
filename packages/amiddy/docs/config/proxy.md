
# proxy
Proxy configuration.


## proxy.options
For complete list of options see [http-proxy#options](https://www.npmjs.com/package/http-proxy#options)

> Note: ssl option is not yet supported.

## proxy.response
Properties for the proxy when response is returned.

## proxy.response.headers
Headers that should be added to the response.

