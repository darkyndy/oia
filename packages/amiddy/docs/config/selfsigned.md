# selfsigned
Options to generate selfsigned certificate.


## selfsigned.attrs
Please refer to: [https://github.com/digitalbazaar/forge/blob/0.7.5/lib/x509.js#L129](https://github.com/digitalbazaar/forge/blob/0.7.5/lib/x509.js#L129)


## selfsigned.opts
Please refer to [selfsigned#options](https://github.com/jfromaniello/selfsigned#options)
