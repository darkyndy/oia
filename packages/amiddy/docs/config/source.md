# source
Source server, usually is your local server.


## source.ip

- Data type: String
- Required: No
- Default value: `127.0.0.1`
- Example value:
    - `192.168.10.2`
    - `127.100.5.24`
- Details:
    - Specify the IP that will be used to request resources if there is no match from dependencies.


## source.https

- Data type: Boolean
- Required: No
- Default value: `false`
- Example value:
    - `false`
    - `true`
- Details:
    - Specify if the source uses secure protocol.


## source.port

- Data type: Number
- Required: No
- Default value: `3000`
- Example value:
    - `80`
    - `8080`
    - `1080`
- Details:
    - Specify port for the source.
