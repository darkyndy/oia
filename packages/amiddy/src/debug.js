
/**
 * Private api for debug
 *
 * @protected
 * @const {Object}
 */
const privateApi = {};

/**
 * Store if we are in debug mode
 *
 * @type {Boolean}
 */
privateApi.isActive = false;


const service = {};

/**
 * Activate debug mode
 */
service.activate = () => {
  privateApi.isActive = true;
};

/**
 * Log message
 *
 * @param {String} msg
 */
service.log = (msg) => {
  privateApi.isActive && console.debug(msg); // eslint-disable-line no-console
};

/**
 * Log multiple messages
 *
 * @param {...String} msgs
 */
service.block = (...msgs) => {
  privateApi.isActive && msgs.forEach(
    (msg) => {
      console.debug(msg); // eslint-disable-line no-console
    }
  );
};


// only for testing
export {privateApi};

export default service;
