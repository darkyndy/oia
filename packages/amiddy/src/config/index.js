import fs from 'fs';
import path from 'path';
import stripComments from 'strip-json-comments';


import debug from '../debug';


const privateApi = {};

/**
 * Synchronously read file contents.
 *
 * @param {String} filePath The filename to read.
 * @return {String} The file contents, with the BOM removed.
 */
privateApi.readFile = (filePath) => (
  fs.readFileSync(filePath, 'utf8').replace(/^\ufeff/u, '')
);

/**
 * Loads a JSON configuration from a file.
 *
 * @param {String} filePath - The filename to load.
 * @return {Object} config - The configuration object from the file.
 * @throws {Error} If the file cannot be read.
 */
privateApi.loadJSONConfigFile = (filePath) => {
  debug.log(`Loading JSON config file: ${filePath}`);

  try {
    return JSON.parse(stripComments(privateApi.readFile(filePath)));
  } catch (e) {
    debug.log(`Error reading JSON file: ${filePath}`);
    e.message = `Cannot read config file: ${filePath}\nError: ${e.message}`;
    e.messageData = {
      message: e.message,
      path: filePath,
    };
    throw e;
  }
};

/**
 * Test if the provided filePath represents a file
 *
 * @param {String} filePath
 * @return {Boolean}
 */
privateApi.isFile = (filePath) => (
  fs.existsSync(filePath) && !fs.lstatSync(filePath).isDirectory()
);

/**
 * Test if the provided filePath represents a file
 *
 * @param {String} pathToResolve
 * @return {String}
 * @throws {Error}
 */
privateApi.getAbsolutePath = (pathToResolve) => {
  const cwd = process.cwd();

  const absolutePath = path.isAbsolute(pathToResolve) ?
    pathToResolve : path.resolve(cwd, pathToResolve);

  if (privateApi.isFile(absolutePath)) {
    return absolutePath;
  }

  throw Error(` Cannot resolve config file: ${pathToResolve}\n cwd: ${cwd}`);
};

/**
 * Test if the provided config object is valid
 *
 * @param {Object} configObj
 * @throws {Error}
 */
privateApi.validate = (configObj) => {
  const vhostConf = configObj.vhost;
  if (!vhostConf) {
    throw Error('Missing `vhost` property');
  }
  if (!vhostConf.name) {
    throw Error('Missing `vhost.name` property');
  }

  const source = configObj.source;
  if (source && (typeof source !== 'object' || Array.isArray(source))) {
    throw Error('`source` property should be an object');
  }

  const deps = configObj.deps;
  if (!deps) {
    throw Error('Missing `deps` property');
  }

  if (!Array.isArray(deps)) {
    throw Error('`deps` property should be an array');
  }

  const depErrors = [];
  deps.forEach(
    (dep, index) => {
      if (!dep.ip && !dep.name) {
        depErrors.push(`Dependency with index ${index} is missing "ip" or "name"`);
      }
      if (!dep.patterns || !Array.isArray(dep.patterns) || dep.patterns.length < 1) {
        depErrors.push(`Dependency with index ${index} should have patterns defined as array of strings`);
      }
    }
  );

  if (depErrors.length > 0) {
    throw Error(depErrors.join(' \n'));
  }
};

/**
 * Get value for port.
 *
 * @param {Boolean} [https=false]
 * @param {Number} [defaultPort]
 * @return {Number} port
 */
privateApi.getPort = (https, defaultPort) => {
  if (defaultPort) {
    return defaultPort;
  }

  return https ? 443 : 80;
};

/**
 * Update config object with defaults for source
 *
 * @param {Object} configObj
 */
privateApi.setSourceDefaults = (configObj) => {
  configObj.source = configObj.source || {};

  configObj.source.ip = configObj.source.ip || '127.0.0.1';

  configObj.source.https = configObj.source.https || false;

  configObj.source.port = configObj.source.port || privateApi.getPort(configObj.source.https, 3000);
};

/**
 * Update config object with defaults for vhost
 *
 * @param {Object} configObj
 */
privateApi.setVhostDefaults = (configObj) => {
  configObj.vhost.https = configObj.vhost.https || false;
  configObj.vhost.port = configObj.vhost.port || privateApi.getPort(configObj.vhost.https);
};

/**
 * Update config object with defaults for every dependency
 *
 * @param {Object} configObj
 */
privateApi.setDepsDefaults = (configObj) => {
  configObj.deps.forEach(
    (dep) => {
      dep.https = dep.https || false;
      dep.port = dep.port || privateApi.getPort(dep.https);
    }
  );
};

/**
 * Update config object with defaults
 *
 * @param {Object} configObj
 */
privateApi.setDefaults = (configObj) => {
  privateApi.setSourceDefaults(configObj);
  privateApi.setVhostDefaults(configObj);
  privateApi.setDepsDefaults(configObj);

};


const service = {};

/**
 * Get configuration. If no path is provided will use `.amiddy` file
 *  that should be at the same level with package.json
 *
 * @param {String} [pathToResolve]
 * @return {Object}
 */
service.get = (pathToResolve) => {
  const pathToUse = pathToResolve || '.amiddy';
  const absolutePath = privateApi.getAbsolutePath(pathToUse);

  const configObj = privateApi.loadJSONConfigFile(absolutePath);

  privateApi.validate(configObj);

  privateApi.setDefaults(configObj);

  return configObj;
};


// only for testing
export {privateApi};

export default service;
