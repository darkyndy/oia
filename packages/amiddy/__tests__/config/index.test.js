
import path from 'path';

// testing file
import config, {privateApi} from '../../src/config';


import debug from '../../src/debug';


// mocks
jest.mock('path', () => (
  {
    isAbsolute: jest.fn().mockReturnValue(true),
    resolve: jest.fn().mockReturnValue('path::resolve'),
  }
));

jest.mock('../../src/debug', () => (
  {
    log: jest.fn(),
  }
));


describe('config', () => {
  let testSpecificMocks;

  beforeEach(() => {
    testSpecificMocks = {};
  });


  describe('privateApi.readFile', () => {

    it('returns content of the file as string when file exists', () => {
      expect(privateApi.readFile('__tests__/__fixtures__/config.json')).toMatchSnapshot();
    });

    it('throws error when file does not exists', () => {
      expect(() => {
        privateApi.readFile('__tests__/__fixtures__/noap.json');
      }).toThrowErrorMatchingSnapshot();
    });

  });

  describe('privateApi.loadJSONConfigFile', () => {

    afterEach(() => {
      debug.log.mockClear();
    });

    it('logs debug message with the file path', () => {
      privateApi.loadJSONConfigFile('__tests__/__fixtures__/config.json');

      expect(debug.log).toHaveBeenCalledWith(
        'Loading JSON config file: __tests__/__fixtures__/config.json'
      );
    });

    it('logs debug message with the file path and error message if there was a problem', () => {
      try {
        privateApi.loadJSONConfigFile('__tests__/__fixtures__/invalid.json');
      } catch (e) {
      }

      expect(
        debug.log.mock.calls
      ).toEqual(
        [
          [
            'Loading JSON config file: __tests__/__fixtures__/invalid.json',
          ],
          [
            'Error reading JSON file: __tests__/__fixtures__/invalid.json',
          ],
        ]
      );
    });

    it('returns content of the file as object when file exists and is valid', () => {
      expect(privateApi.loadJSONConfigFile('__tests__/__fixtures__/config.json')).toMatchSnapshot();
    });

    it('throws error when file exists but does not have valid json', () => {
      expect(() => {
        privateApi.loadJSONConfigFile('__tests__/__fixtures__/invalid.json');
      }).toThrowErrorMatchingSnapshot();
    });

    it('throws error when file does not exists', () => {
      expect(() => {
        privateApi.loadJSONConfigFile('__tests__/__fixtures__/noap.json');
      }).toThrowErrorMatchingSnapshot();
    });

  });

  describe('privateApi.isFile', () => {

    it('returns true when provided path as string represents a file', () => {
      expect(privateApi.isFile('__tests__/__fixtures__/config.json')).toBe(true);
    });

    it('returns false when provided path as string is a folder', () => {
      expect(privateApi.isFile('__tests__/__fixtures__')).toBe(false);
    });

    it('returns false when provided path as string is invalid', () => {
      expect(privateApi.isFile('__tests__/__fixtures__/noap.json')).toBe(false);
    });

  });

  describe('privateApi.getAbsolutePath', () => {
    beforeAll(() => {
      jest.spyOn(process, 'cwd').mockReturnValue('process::cwd');
      jest.spyOn(privateApi, 'isFile').mockReturnValue(true);
    });
    beforeEach(() => {
      testSpecificMocks.pathToResolve = 'path/to/resolve';
    });

    afterEach(() => {
      path.isAbsolute.mockClear();
      path.resolve.mockClear();
      process.cwd.mockClear();
      privateApi.isFile.mockClear();
    });
    afterAll(() => {
      process.cwd.mockRestore();
      privateApi.isFile.mockRestore();
    });

    it('determines current working directory of the Node.js process', () => {
      privateApi.getAbsolutePath(testSpecificMocks.pathToResolve);

      expect(
        process.cwd
      ).toHaveBeenCalledWith();
    });

    it('determines if provided path is absolute', () => {
      privateApi.getAbsolutePath(testSpecificMocks.pathToResolve);

      expect(
        path.isAbsolute
      ).toHaveBeenCalledWith(
        testSpecificMocks.pathToResolve
      );
    });

    it('tries to resolve path when is not absolute', () => {
      path.isAbsolute.mockReturnValueOnce(false);
      privateApi.getAbsolutePath(testSpecificMocks.pathToResolve);

      expect(
        path.resolve
      ).toHaveBeenCalledWith(
        'process::cwd',
        testSpecificMocks.pathToResolve,
      );
    });

    it('determines if we are working with a file based on absolute path (path is absolute)', () => {
      privateApi.getAbsolutePath(testSpecificMocks.pathToResolve);

      expect(
        privateApi.isFile
      ).toHaveBeenCalledWith(
        testSpecificMocks.pathToResolve,
      );
    });

    it('determines if we are working with a file based on absolute path (path is relative)', () => {
      path.isAbsolute.mockReturnValueOnce(false);
      privateApi.getAbsolutePath(testSpecificMocks.pathToResolve);

      expect(
        privateApi.isFile
      ).toHaveBeenCalledWith(
        'path::resolve',
      );
    });

    it('returns absolute path when it represents a file (path is absolute)', () => {
      privateApi.getAbsolutePath(testSpecificMocks.pathToResolve);

      expect(
        privateApi.isFile
      ).toHaveBeenCalledWith(
        testSpecificMocks.pathToResolve,
      );
    });

    it('returns absolute path when it represents a file (path is relative)', () => {
      path.isAbsolute.mockReturnValueOnce(false);
      privateApi.getAbsolutePath(testSpecificMocks.pathToResolve);

      expect(
        privateApi.isFile
      ).toHaveBeenCalledWith(
        'path::resolve',
      );
    });

    it('throws error when the path does not represents a file', () => {
      privateApi.isFile.mockReturnValueOnce(false);

      expect(
        () => {
          privateApi.getAbsolutePath(testSpecificMocks.pathToResolve);
        }
      ).toThrowErrorMatchingSnapshot();
    });

  });

  describe('privateApi.validate', () => {
    beforeEach(() => {
      testSpecificMocks.configObj = {
        deps: [
          {
            name: '127.0.0.2',
            patterns: [
              '/images/**',
            ],
            port: 3000,
          },
          {
            name: '127.0.0.1',
            patterns: [
              '/api/**',
            ],
            port: 8080,
          },
        ],
        proxy: {
          ws: true,
        },
        source: {
          name: '127.0.0.1',
          port: 3000,
        },
        vhost: {
          name: 'http://github.com',
          port: 80,
        },
      };
    });

    it('no error is thrown when config object is valid', () => {
      expect(
        () => {
          privateApi.validate(testSpecificMocks.configObj);
        }
      ).not.toThrow();
    });

    it('throws error when missing `vhost`', () => {
      testSpecificMocks.configObj.vhost = undefined;

      expect(
        () => {
          privateApi.validate(testSpecificMocks.configObj);
        }
      ).toThrowErrorMatchingSnapshot();
    });

    it('throws error when missing `vhost.name`', () => {
      testSpecificMocks.configObj.vhost.name = undefined;

      expect(
        () => {
          privateApi.validate(testSpecificMocks.configObj);
        }
      ).toThrowErrorMatchingSnapshot();
    });

    it('throws error when `source` is not an object (string)', () => {
      testSpecificMocks.configObj.source = 'str';

      expect(
        () => {
          privateApi.validate(testSpecificMocks.configObj);
        }
      ).toThrowErrorMatchingSnapshot();
    });

    it('throws error when `source` is not an object (array)', () => {
      testSpecificMocks.configObj.source = [];

      expect(
        () => {
          privateApi.validate(testSpecificMocks.configObj);
        }
      ).toThrowErrorMatchingSnapshot();
    });

    it('throws error when missing `deps`', () => {
      testSpecificMocks.configObj.deps = undefined;

      expect(
        () => {
          privateApi.validate(testSpecificMocks.configObj);
        }
      ).toThrowErrorMatchingSnapshot();
    });

    it('throws error when `deps` is not an array', () => {
      testSpecificMocks.configObj.deps = {};

      expect(
        () => {
          privateApi.validate(testSpecificMocks.configObj);
        }
      ).toThrowErrorMatchingSnapshot();
    });

    it('throws error for every `dependency` that does not have patterns as array and ip or name', () => {
      testSpecificMocks.configObj.deps = [
        {
          ip: '127.0.0.1',
          name: 'example.com',
          patterns: [
            'aaaa',
          ],
        },
        {
          patterns: [],
        },
        {
          patterns: [
            'aaaa',
          ],
        },
        {
          ip: '127.0.0.1',
          patterns: [
            'aaaa',
          ],
        },
        {
          name: 'example.com',
          patterns: [
            'aaaa',
          ],
        },
      ];

      expect(
        () => {
          privateApi.validate(testSpecificMocks.configObj);
        }
      ).toThrowErrorMatchingSnapshot();
    });

  });

  describe('privateApi.getPort', () => {
    beforeEach(() => {
      testSpecificMocks.https = true;
      testSpecificMocks.defaultPort = 3000;
    });

    it('returns default port provided when it has truthy value', () => {
      expect(
        privateApi.getPort(testSpecificMocks.https, testSpecificMocks.defaultPort)
      ).toBe(3000);
    });

    it('returns 443 when default port has falsy value and https is truthy', () => {
      expect(
        privateApi.getPort(testSpecificMocks.https)
      ).toBe(443);
    });

    it('returns 80 when default port has falsy value and https is falsy', () => {
      testSpecificMocks.https = false;

      expect(
        privateApi.getPort(testSpecificMocks.https)
      ).toBe(80);
    });
  });

  describe('privateApi.setSourceDefaults', () => {
    beforeAll(() => {
      jest.spyOn(privateApi, 'getPort').mockReturnValue(80);
    });
    beforeEach(() => {
      testSpecificMocks.configObj = {
        deps: [],
        vhost: {},
      };
    });

    afterEach(() => {
      privateApi.getPort.mockClear();
    });
    afterAll(() => {
      privateApi.getPort.mockRestore();
    });

    it('when source port has falsy value it prepares port based on source.https and default port as 3000', () => {
      testSpecificMocks.configObj.source = {
        https: true,
      };
      privateApi.setSourceDefaults(testSpecificMocks.configObj);

      expect(
        privateApi.getPort
      ).toHaveBeenCalledWith(
        testSpecificMocks.configObj.source.https,
        3000,
      );
    });

    it('updates by reference config object with source data (all props)', () => {
      privateApi.setSourceDefaults(testSpecificMocks.configObj);

      expect(
        testSpecificMocks.configObj
      ).toStrictEqual(
        {
          deps: [],
          source: {
            https: false,
            ip: '127.0.0.1',
            port: privateApi.getPort(),
          },
          vhost: {},
        }
      );
    });

    it('updates by reference config object with source data (missing props)', () => {
      testSpecificMocks.configObj.source = {
        port: 1098,
      };
      privateApi.setSourceDefaults(testSpecificMocks.configObj);

      expect(
        testSpecificMocks.configObj
      ).toStrictEqual(
        {
          deps: [],
          source: {
            https: false,
            ip: '127.0.0.1',
            port: 1098,
          },
          vhost: {},
        }
      );
    });

  });

  describe('privateApi.setVhostDefaults', () => {
    beforeAll(() => {
      jest.spyOn(privateApi, 'getPort').mockReturnValue(80);
    });
    beforeEach(() => {
      testSpecificMocks.configObj = {
        deps: [],
        source: {},
        vhost: {
          name: 'example.com',
        },
      };
    });

    afterEach(() => {
      privateApi.getPort.mockClear();
    });
    afterAll(() => {
      privateApi.getPort.mockRestore();
    });

    it('when vhost port has falsy value it prepares port based on vhost.https', () => {
      testSpecificMocks.configObj.vhost.https = true;
      privateApi.setVhostDefaults(testSpecificMocks.configObj);

      expect(
        privateApi.getPort
      ).toHaveBeenCalledWith(
        testSpecificMocks.configObj.vhost.https,
      );
    });

    it('updates by reference config object with vhost data (all props)', () => {
      privateApi.setVhostDefaults(testSpecificMocks.configObj);

      expect(
        testSpecificMocks.configObj
      ).toStrictEqual(
        {
          deps: [],
          source: {},
          vhost: {
            https: false,
            name: 'example.com',
            port: privateApi.getPort(),
          },
        }
      );
    });

    it('updates by reference config object with vhost data (missing props)', () => {
      testSpecificMocks.configObj.vhost.port = 1098;
      privateApi.setVhostDefaults(testSpecificMocks.configObj);

      expect(
        testSpecificMocks.configObj
      ).toStrictEqual(
        {
          deps: [],
          source: {},
          vhost: {
            https: false,
            name: 'example.com',
            port: 1098,
          },
        }
      );
    });

  });

  describe('privateApi.setDepsDefaults', () => {
    beforeAll(() => {
      jest.spyOn(privateApi, 'getPort').mockReturnValue(80);
    });
    beforeEach(() => {
      testSpecificMocks.configObj = {
        deps: [
          {
            id: '8.8.7.7',
            patterns: [
              '/v1',
            ],
          },
          {
            https: true,
            name: 'example.com',
            patterns: [
              '/v2',
            ],
          },
        ],
        source: {},
        vhost: {},
      };
    });

    afterEach(() => {
      privateApi.getPort.mockClear();
    });
    afterAll(() => {
      privateApi.getPort.mockRestore();
    });

    it('when dependency port has falsy value it prepares port based on dependency.https', () => {
      testSpecificMocks.configObj.vhost.https = true;
      privateApi.setDepsDefaults(testSpecificMocks.configObj);

      expect(
        privateApi.getPort.mock.calls
      ).toEqual(
        [
          [
            false,
          ],
          [
            true,
          ],
        ]
      );
    });

    it('updates by reference config object with deps data', () => {
      privateApi.setDepsDefaults(testSpecificMocks.configObj);

      expect(
        testSpecificMocks.configObj
      ).toStrictEqual(
        {
          deps: [
            {
              https: false,
              id: '8.8.7.7',
              patterns: [
                '/v1',
              ],
              port: privateApi.getPort(),
            },
            {
              https: true,
              name: 'example.com',
              patterns: [
                '/v2',
              ],
              port: privateApi.getPort(),
            },
          ],
          source: {},
          vhost: {},
        }
      );
    });

  });

  describe('privateApi.setDefaults', () => {
    beforeAll(() => {
      jest.spyOn(privateApi, 'setSourceDefaults').mockReturnValue(undefined);
      jest.spyOn(privateApi, 'setVhostDefaults').mockReturnValue(undefined);
      jest.spyOn(privateApi, 'setDepsDefaults').mockReturnValue(undefined);
    });
    beforeEach(() => {
      testSpecificMocks.configObj = {
        proxy: {
          ws: true,
        },
        source: {
          name: '127.0.0.1',
        },
        vhost: {
          name: 'http://github.com',
        },
      };
    });

    afterEach(() => {
      privateApi.setSourceDefaults.mockClear();
      privateApi.setVhostDefaults.mockClear();
      privateApi.setDepsDefaults.mockClear();
    });
    afterAll(() => {
      privateApi.setSourceDefaults.mockRestore();
      privateApi.setVhostDefaults.mockRestore();
      privateApi.setDepsDefaults.mockRestore();
    });

    it('updates by reference source from config obj', () => {
      privateApi.setDefaults(testSpecificMocks.configObj);

      expect(
        privateApi.setSourceDefaults
      ).toHaveBeenCalledWith(
        testSpecificMocks.configObj
      );
    });

    it('updates by reference vhost from config obj', () => {
      privateApi.setDefaults(testSpecificMocks.configObj);

      expect(
        privateApi.setVhostDefaults
      ).toHaveBeenCalledWith(
        testSpecificMocks.configObj
      );
    });

    it('updates by reference deps from config obj', () => {
      privateApi.setDefaults(testSpecificMocks.configObj);

      expect(
        privateApi.setDepsDefaults
      ).toHaveBeenCalledWith(
        testSpecificMocks.configObj
      );
    });

  });

  describe('get', () => {
    beforeAll(() => {
      jest.spyOn(privateApi, 'setDefaults').mockReturnValue(undefined);
      jest.spyOn(privateApi, 'validate').mockReturnValue(undefined);
      jest.spyOn(privateApi, 'getAbsolutePath').mockReturnValue('/absolute/path/to/file');
      jest.spyOn(privateApi, 'loadJSONConfigFile').mockReturnValue(
        {
          data: 'confiObj',
        }
      );
    });
    beforeEach(() => {
      testSpecificMocks.pathToResolve = 'path/to/resolve';
    });

    afterEach(() => {
      privateApi.setDefaults.mockClear();
      privateApi.validate.mockClear();
      privateApi.getAbsolutePath.mockClear();
      privateApi.loadJSONConfigFile.mockClear();
    });
    afterAll(() => {
      privateApi.setDefaults.mockRestore();
      privateApi.validate.mockRestore();
      privateApi.getAbsolutePath.mockRestore();
      privateApi.loadJSONConfigFile.mockRestore();
    });

    it('uses `.amiddy` as file when path has falsy value', () => {
      config.get();

      expect(
        privateApi.getAbsolutePath
      ).toHaveBeenCalledWith(
        '.amiddy',
      );
    });

    it('retrieves absolute path to file', () => {
      config.get(testSpecificMocks.pathToResolve);

      expect(
        privateApi.getAbsolutePath
      ).toHaveBeenCalledWith(
        testSpecificMocks.pathToResolve,
      );
    });

    it('loads json file', () => {
      config.get(testSpecificMocks.pathToResolve);

      expect(
        privateApi.loadJSONConfigFile
      ).toHaveBeenCalledWith(
        privateApi.getAbsolutePath(),
      );
    });

    it('validates config object', () => {
      config.get(testSpecificMocks.pathToResolve);

      expect(
        privateApi.validate
      ).toHaveBeenCalledWith(
        {
          data: 'confiObj',
        },
      );
    });

    it('set defaults on config object', () => {
      config.get(testSpecificMocks.pathToResolve);

      expect(
        privateApi.setDefaults
      ).toHaveBeenCalledWith(
        {
          data: 'confiObj',
        },
      );
    });

    it('returns config object if is valid', () => {
      expect(
        config.get(testSpecificMocks.pathToResolve)
      ).toEqual(
        {
          data: 'confiObj',
        },
      );
    });

    it('throws error if config object is not valid', () => {
      privateApi.validate.mockImplementationOnce(
        () => {
          throw 'Throw';
        }
      );

      expect(
        () => {
          config.get(testSpecificMocks.pathToResolve);
        }
      ).toThrow('Throw');
    });

  });

});
