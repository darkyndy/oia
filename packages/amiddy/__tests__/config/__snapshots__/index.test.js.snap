// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config privateApi.getAbsolutePath throws error when the path does not represents a file 1`] = `
" Cannot resolve config file: path/to/resolve
 cwd: process::cwd"
`;

exports[`config privateApi.loadJSONConfigFile returns content of the file as object when file exists and is valid 1`] = `
Object {
  "deps": Array [
    Object {
      "name": "127.0.0.2",
      "patterns": Array [
        "/images/**",
      ],
      "port": 3000,
    },
    Object {
      "name": "127.0.0.1",
      "patterns": Array [
        "/api/**",
      ],
      "port": 8080,
    },
  ],
  "proxy": Object {
    "ws": true,
  },
  "source": Object {
    "name": "127.0.0.1",
    "port": 3000,
  },
  "vhost": Object {
    "name": "http://github.com",
    "port": 80,
  },
}
`;

exports[`config privateApi.loadJSONConfigFile throws error when file does not exists 1`] = `
"Cannot read config file: __tests__/__fixtures__/noap.json
Error: ENOENT: no such file or directory, open '__tests__/__fixtures__/noap.json'"
`;

exports[`config privateApi.loadJSONConfigFile throws error when file exists but does not have valid json 1`] = `
"Cannot read config file: __tests__/__fixtures__/invalid.json
Error: Unexpected token i in JSON at position 4"
`;

exports[`config privateApi.readFile returns content of the file as string when file exists 1`] = `
"{
  \\"deps\\": [
    {
      \\"name\\": \\"127.0.0.2\\",
      \\"patterns\\": [
        \\"/images/**\\"
      ],
      \\"port\\": 3000
    },
    {
      \\"name\\": \\"127.0.0.1\\",
      \\"patterns\\": [
        \\"/api/**\\"
      ],
      \\"port\\": 8080
    }
  ],
  \\"proxy\\": {
    \\"ws\\": true
  },
  \\"source\\": {
    \\"name\\": \\"127.0.0.1\\",
    \\"port\\": 3000
  },
  \\"vhost\\": {
    \\"name\\": \\"http://github.com\\",
    \\"port\\": 80
  }
}"
`;

exports[`config privateApi.readFile throws error when file does not exists 1`] = `"ENOENT: no such file or directory, open '__tests__/__fixtures__/noap.json'"`;

exports[`config privateApi.validate throws error for every \`dependency\` that does not have patterns as array and ip or name 1`] = `
"Dependency with index 1 is missing \\"ip\\" or \\"name\\" 
Dependency with index 1 should have patterns defined as array of strings 
Dependency with index 2 is missing \\"ip\\" or \\"name\\""
`;

exports[`config privateApi.validate throws error when \`deps\` is not an array 1`] = `"\`deps\` property should be an array"`;

exports[`config privateApi.validate throws error when \`source\` is not an object (array) 1`] = `"\`source\` property should be an object"`;

exports[`config privateApi.validate throws error when \`source\` is not an object (string) 1`] = `"\`source\` property should be an object"`;

exports[`config privateApi.validate throws error when missing \`deps\` 1`] = `"Missing \`deps\` property"`;

exports[`config privateApi.validate throws error when missing \`vhost.name\` 1`] = `"Missing \`vhost.name\` property"`;

exports[`config privateApi.validate throws error when missing \`vhost\` 1`] = `"Missing \`vhost\` property"`;
